expression/builtin_time.go

```diff
@@ -2981,78 +3247,346 @@ type subTimeFunctionClass struct {
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_subtime
+func (b *builtinSubDatetimeAndStringSig) evalTime(row []types.Datum) (types.Time, bool, error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	arg0, isNull, err := b.args[0].EvalTime(row, sc)
+	if isNull || err != nil {
+		return types.ZeroDatetime, isNull, errors.Trace(err)
 	}
-	switch args[0].Kind() {
-	case types.KindMysqlTime:
-		arg0 := args[0].GetMysqlTime()
-		arg1time, err := arg1.ConvertToTime(uint8(getFsp(arg1.String())))
-		if err != nil {
-			return d, errors.Trace(err)
-		}
-		tmpDuration := arg0.Sub(&arg1time)
-		result, err := tmpDuration.ConvertToTime(arg0.Type)
-		if err != nil {
-			return d, errors.Trace(err)
-		}
-		d.SetMysqlTime(result)
-	case types.KindMysqlDuration:
-		arg0 := args[0].GetMysqlDuration()
-		result, err := arg0.Sub(arg1)
+	s, isNull, err := b.args[1].EvalString(row, sc)
+	if isNull || err != nil {
+		return types.ZeroDatetime, isNull, errors.Trace(err)
+	}
+	if err != nil {
+		return types.ZeroDatetime, true, errors.Trace(err)
+	}
+	if !isDuration(s) {
+		return types.ZeroDatetime, true, nil
+	}
+	arg1, err := types.ParseDuration(s, getFsp(s))
+	if err != nil {
+		return types.ZeroDatetime, true, errors.Trace(err)
+	}
+	arg1time, err := arg1.ConvertToTime(mysql.TypeDatetime)
+	if err != nil {
+		return types.ZeroDatetime, true, errors.Trace(err)
+	}
+	tmpDuration := arg0.Sub(&arg1time)
+	result, err := tmpDuration.ConvertToTime(mysql.TypeDatetime)
+	if err != nil {
+		return result, true, errors.Trace(err)
+	}
+	return result, false, nil
+}
+
+type builtinSubTimeDateTimeNullSig struct {
+	baseTimeBuiltinFunc
+}
+
+// evalTime evals a builtinSubTimeDateTimeNullSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_subtime
+func (b *builtinSubTimeDateTimeNullSig) evalTime(row []types.Datum) (types.Time, bool, error) {
+	return types.ZeroDatetime, true, nil
+}
+
+type builtinSubStringAndDurationSig struct {
+	baseStringBuiltinFunc
+}
+
+// evalString eval a builtinSubStringAndDurationSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_subtime
+func (b *builtinSubStringAndDurationSig) evalString(row []types.Datum) (result string, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	var (
+		ss   string
+		arg1 types.Duration
+	)
+	ss, isNull, err = b.args[0].EvalString(row, sc)
+	if isNull || err != nil {
+		return "", isNull, errors.Trace(err)
+	}
+	arg1, isNull, err = b.args[1].EvalDuration(row, sc)
+	if isNull || err != nil {
+		return "", isNull, errors.Trace(err)
+	}
+	if isDuration(ss) {
+		result, err = strDurationSubDuration(ss, arg1)
 		if err != nil {
-			return d, errors.Trace(err)
+			return "", true, errors.Trace(err)
 		}
-		d.SetMysqlDuration(result)
-	default:
-		ss, err := args[0].ToString()
+		return result, false, nil
+	}
+	result, err = strDatetimeSubDuration(ss, arg1)
+	if err != nil {
```

Comment:
remove this check
return result, err!=nil, errors.Trace(err)

Reply:
ok